probes: replace ProbeManager with features API in cilium/ebpf

rgo3 opened the pull request:
The first commit makes sure we call `rlimit.RemoveMemlock()` only once and early enough on the startup of the agent for all probes being able to execute without `EPERM`.
 
The second commit vendors the github.com/cilium/ebpf/features package.

With the third commit we replace most of the feature probes done with the current ProbeManager system with the features API in cilium/ebpf.

Fixes: #20062

rgo3 committed:
probes: replace GetMapTypes, GetHelpers() and GetMisc() in ProbeManager

To replace the usage of bpftool to do feature checks in cilium
this commit replaces calls to the ProbeManager with
calls to the features API in github.com/cilium/ebpf.

GetMapTypes(), GetHelpers() and GetMisc() were fully replaced and removed.

Signed-off-by: Robin Gögge <r.goegge@isovalent.com>

ti-mo commented on pkg/datapath/linux/devices.go:
```diff
@@ -497,10 +499,8 @@ func findK8SNodeIPLink() (netlink.Link, error) {
-	if h := probesManager.GetHelpers("sched_cls"); h != nil {
-		_, found := h["bpf_skb_change_head"]
-		return found
+	if err := features.HaveProgramHelper(ebpf.SchedCLS, asm.FnSkbChangeHead); err == nil {
```
Replace the whole function with
```suggestion
	return features.HaveProgramHelper(ebpf.SchedCLS, asm.FnSkbChangeHead) == nil
```

ti-mo commented on pkg/datapath/linux/requirements.go:
```diff
@@ -138,8 +139,7 @@ func CheckMinRequirements() {
 		if option.Config.EnableVTEP {
-			supportedMisc := probeManager.GetMisc()
-			if !supportedMisc.HaveLargeInsnLimit {
+			if err := features.HaveLargeInstructions(); err != nil {
```
`err` is unused outside of the nil check:
```suggestion
			if features.HaveLargeInstructions() != nil {
```

ti-mo commented on pkg/redirectpolicy/manager.go:
```diff
@@ -104,15 +106,9 @@ func (rpm *Manager) RegisterGetStores(sg StoreGetter) {
-			}
-		}
-		if !found {
+		lookupTCPErr := features.HaveProgramHelper(ebpf.CGroupSockAddr, asm.FnSkLookupTcp)
```
No need to put these in a variable.

ti-mo commented on pkg/datapath/linux/probes/probes.go:
```diff
@@ -312,31 +310,11 @@ func (p *ProbeManager) GetOptionalConfig() map[KernelParam]kernelOption {
 }
 
 // GetMapTypes returns information about supported BPF map types.
+// Deprecated. Use github.com/cilium/ebpf/features instead.
```
```suggestion
// Deprecated. Use github.com/cilium/ebpf/features.HaveMapType() instead.
```

ti-mo commented on pkg/bandwidth/bandwidth.go:
```diff
@@ -41,14 +43,13 @@ func ProbeBandwidthManager() {
 	}
 
 	kernelGood := false
```
```suggestion
	kernelGood := features.HaveProgramHelper(ebpf.SchedCLS, asm.FnSkbEcnSetCe) == nil
```

ti-mo commented on daemon/cmd/kube_proxy_replacement.go:
```diff
@@ -580,9 +530,12 @@ func finishKubeProxyReplacementInit(isKubeProxyReplacementStrict bool) error {
 	option.Config.NodePortNat46X64 = option.Config.EnableIPv4 && option.Config.EnableIPv6 &&
-		option.Config.NodePortMode == option.NodePortModeSNAT &&
-		probes.NewProbeManager().GetMisc().HaveLargeInsnLimit
+		option.Config.NodePortMode == option.NodePortModeSNAT && haveLargeInsnLimit
```
```suggestion
		option.Config.NodePortMode == option.NodePortModeSNAT && 
		features.HaveLargeInstructions() == nil
```

ti-mo reviewed (changes requested):
:very-nice:

Most all `if err := features.HaveXYZ(); err ...` instances can be replaced by `if features.Have... {` if `err` is only used in the conditional.

I've highlighted a few such cases, but please apply throughout.

:rocket:

borkmann commented on pkg/datapath/connector/ipvlan.go:
```diff
@@ -225,8 +225,13 @@ func CreateAndSetupIpvlanSlave(id string, slaveIfName string, netNs ns.NetNS, mt
+	}
+
+	mapID, ok := mapInfo.ID()
+	if !ok {
```
Needs rebase, then this commit can be dropped since we removed ipvlan support.

borkmann commented on daemon/cmd/daemon_main.go:
```diff
@@ -1534,20 +1537,17 @@ func initEnv(cmd *cobra.Command) {
-				option.Config.EnableBPFTProxy = false
-				log.Info("Disabled support for BPF TProxy due to missing kernel support for socket assign (Linux 5.7 or later)")
-			}
+		if err := features.HaveProgramHelper(ebpf.SchedACT, asm.FnSkAssign); err != nil {
```
Small nit: I would use SchedCLS as well here, since we don't use act_bpf internally, so we also don't need to add unnecessary dependency. (Both enable the same helpers.)

borkmann commented on daemon/cmd/datapath.go:
```diff
@@ -352,10 +353,9 @@ func (d *Daemon) initMaps() error {
+	lruHashErr := features.HaveMapType(ebpf.LRUHash)
 	createSockRevNatMaps := option.Config.EnableSocketLB &&
-		option.Config.EnableHostServicesUDP && supportedMapTypes.HaveLruHashMapType
+		option.Config.EnableHostServicesUDP && lruHashErr == nil
```
nit: Do we need extra var? maybe just fold it in: `&& features.HaveMapType(ebpf.LRUHash) == nil`?

borkmann commented on daemon/cmd/kube_proxy_replacement.go:
```diff
@@ -216,13 +215,7 @@ func initKubeProxyReplacementOptions() (bool, error) {
-			}
-		}
-		if !found {
+		if err := features.HaveProgramHelper(ebpf.SchedACT, asm.FnFibLookup); err != nil {
```
nit: independent of your PR, but same comment: SchedCLS

borkmann commented on pkg/datapath/loader/compile.go:
```diff
@@ -157,22 +159,20 @@ var (
-			if manager.GetMisc().HaveV3ISAExtension && h != nil {
-				if _, ok := h["bpf_redirect_neigh"]; ok {
+			if err := features.HaveV3ISA(); err == nil {
+				if err := features.HaveProgramHelper(ebpf.SchedCLS, asm.FnRedirectNeigh); err == nil {
```
nit: err var unused

borkmann commented on pkg/datapath/loader/compile.go:
```diff
@@ -157,22 +159,20 @@ var (
 			// We want to enable v2 on all kernels that support it, that is,
 			// kernels 4.14+.
-			if manager.GetMisc().HaveV2ISAExtension {
+			if err := features.HaveV2ISA(); err == nil {
```
nit: ditto

rgo3 committed:
vendor: add github.com/cilium/ebpf/features

Signed-off-by: Robin Gögge <r.goegge@isovalent.com>

rgo3 committed:
daemon: only set rlimit Memlock once

This commit removes all calls to rlimit.RemoveMemlock()
in favor of calling it once when the agent starts up.

Signed-off-by: Robin Gögge <r.goegge@isovalent.com>

joestringer commented on daemon/cmd/daemon_main.go:
```diff
@@ -1478,20 +1480,17 @@ func initEnv(cmd *cobra.Command) {
 				log.Info("Disabled support for IPv4 fragments due to missing kernel support for BPF LRU maps")
 			}
 		}
 	}
 
 	if option.Config.EnableBPFTProxy {
-		if h := probes.NewProbeManager().GetHelpers("sched_act"); h != nil {
-			if _, ok := h["bpf_sk_assign"]; !ok {
-				option.Config.EnableBPFTProxy = false
-				log.Info("Disabled support for BPF TProxy due to missing kernel support for socket assign (Linux 5.7 or later)")
-			}
+		if probes.HaveProgramHelper(ebpf.SchedCLS, asm.FnSkAssign) != nil {
```
The previous code says `sched_act`, but the new code says `SchedCLS`. I don't think these are the same thing?

joestringer replied:
(ditto for the rest of this commit, I won't repeat the comment in each location)

joestringer replied:
Hmm, the more I look through the PR, the more it seems like the existing code was just inconsistently checking `sched_act` vs. `sched_cls`. From memory I think we actually use `sched_act` (direct action) for attachment rather than attaching a classifier, but I could be wrong.

ti-mo replied:
`sched_act` and `sched_cls` were introduced in the same commit and have been in the kernel almost since the start of ebpf. `sk_assign` can be used in both cls and act, so the probe result here should be equivalent. Not sure why this was changed, but at least now we're consistently using `SchedCLS` as you pointed out! :wink:

joestringer commented on pkg/datapath/linux/devices.go:
```diff
@@ -501,12 +503,7 @@ func findK8SNodeIPLink() (netlink.Link, error) {
-		return found
-	}
-	return false
+	return probes.HaveProgramHelper(ebpf.SchedCLS, asm.FnSkbChangeHead) == nil
```
Incidentally, it seems like a curious choice to expose the functions via the `asm` package rather than just in the ebpf package, since the asm defines the BPF_CALL instruction but the interpretation of the arguments (ie numbers here) is up to the eBPF runtime. That is to say, depending on which platform we target, the actual meaning of the number 5 which is currently mapped to some specific function in the asm package could be different. But that's probably more feedback for ensuring the ebpf library is cross-platform rather than this specific PR. (Though that said, if we change this in the upstream library then obviously this PR would also need updating).

joestringer replied:
^^ I'm not quite sure what this means in the end. Maybe creating a `github.com/cilium/ebpf/linux` package specifically to define these constants for the Linux eBPF runtime?

I'll cc @dthaler here in case he has thoughts / opinions on this, given his background on the eBPF for Windows project.

ti-mo replied:
> Incidentally, it seems like a curious choice to expose the functions via the asm package rather than just in the ebpf package ...

That sounds like a strawman argument to me. The `asm` package is where the ISA description lives and is used by the instruction decoder. When it was conceived, it made sense to put these constants here, since Linux had the only real eBPF implementation at the time (ignoring ubpf for a second), and why would anyone ever want to write another runtime with different helpers? :wink:

Since helpers are part of the ABI and we were only targeting Linux, they were implicitly considered part of the ISA. Until the `features` package came around, helper type constants weren't taken by any APIs outside of the `asm` package, so they were really values you'd only use when constructing bytecode from Go. It may look a bit off-color, but there's still not much use for them elsewhere in the API, so IMO they're fine where they are.

We could consider putting these behind build tags, but the existing ones will need to remain available in `asm` for compatibility. I don't see a real need to change this until someone comes along to add Windows support, though.

joestringer replied:
I raise this question specifically because of the active discussions around cross-platform support going on at the moment in the context of the eBPF for Windows project. Cross-platform eBPF in general was discussed as recently as two weeks ago as part of the eBPF community call ([notes](https://lore.kernel.org/bpf/CH2PR21MB14649A069BDEACFBC440C623A3649@CH2PR21MB1464.namprd21.prod.outlook.com/T/#u)). One of the points highlighted from recent discussions is that the helper codes for that instruction are not guaranteed to be consistent between platforms. When you look at the current helpers, some of those helpers are very Linux specific, so it will not make sense for other platforms to support the same numbers for those helpers. From what I understand, the current thinking from other runtimes is that they will not ensure ABI compatibility with the Linux runtime for the `BPF_CALL` instruction.

From your explanation, I think there is a completely reasonable line of thinking that leads to the current layout of these. That said, if the question is as simple as which package these symbols live in, I think it's worth at least considering how a future of cilium/ebpf against other runtimes might work so we can minimize any library change impacts in future. Maybe the future answer to this should be as simple as setting some Go build flags when using the eBPF library in order to target specific runtimes. In that case, no further changes would be necessary. I floated the idea above that these ABI symbols could be declared in a separate package, that's also a possibility but I don't feel strongly that this should be the case.

joestringer commented on pkg/datapath/linux/probes/probes.go:
```diff
@@ -422,3 +395,101 @@ func (p *ProbeManager) KernelConfigAvailable() bool {
+
+// HaveMapType is a wrapper around features.HaveMapType() to check if a certain
+// BPF map type is supported by the kernel.
+// On unexpected probe results this function will terminate with log.Fatal().
+func HaveMapType(mt ebpf.MapType) error {
```
I think that `MustHaveMapType()` would be a more canonical way to call this if there are strict consequences for the map type being missing.

joestringer replied:
(next feedback will render this unnecessary)

joestringer replied:
Somehow I just missed the whole `ErrNotSupported` logic, sorry about that. This feedback doesn't really make sense.

joestringer commented on pkg/datapath/linux/probes/probes.go:
```diff
@@ -422,3 +395,101 @@ func (p *ProbeManager) KernelConfigAvailable() bool {
+		return err
+	}
+	if err != nil {
+		log.WithError(err).Fatalf("failed to probe MapType: %s", mt.String())
```
I don't particularly like combining the check for the map type with the forced exit here. For instance, we would know that the map type is missing. But we don't know _why_ that map type is needed. Maybe there's a workaround that involves disabling some feature flag. Only the caller has that context. If we introduce a new map type tomorrow that can be conditionally used depending on which kernel Cilium runs on, then this also prevents us from having an implementation that conditionally checks for that support and falls back to an older / more naive implementation. This is the case today for LRU and LPM map types, for what it's worth. I anticipate we'll need something like this in future too.

I would suggest creating some canonical error here, so above declare a `ErrorMapTypeUnsupported = fmt.Errorf("map type unsupported")` and then here, do:

```suggestion
        return fmt.Errorf("failed to probe MapType %s: %w", mt.String(), ErrorMapTypeUnsupported)
```

Then, the calling code could do `if errors.Is(err, probes.ErrorMapTypeUnsupported) { ... fallback / exit ... }`.

joestringer replied:
(I realize now that `ebpf` library already has the unsupported error so we probably don't need to redefine it)

joestringer replied:
ditto, I just missed the `ErrNotSupported` logic. More :coffee: required.

joestringer commented on pkg/datapath/linux/probes/probes.go:
```diff
@@ -422,3 +395,101 @@ func (p *ProbeManager) KernelConfigAvailable() bool {
+		return err
+	}
+	if err != nil {
+		log.WithError(err).Fatalf("failed to probe ProgramType: %s", pt.String())
```
Again, we can't just fatal out here. SkMsg is only supported in 4.19+ IIRC, but we will support older kernel versions for a while yet. This should return another declared error type.

joestringer replied:
Ditto.

joestringer commented on pkg/datapath/linux/probes/probes.go:
```diff
@@ -422,3 +395,101 @@ func (p *ProbeManager) KernelConfigAvailable() bool {
+		return err
+	}
+	if err != nil {
+		log.WithError(err).Fatalf("failed to probe Helper: %s/%s", pt.String(), helper.String())
```
Same here, we support a range of kernels that have different helpers and sometimes we can fall back to using older helpers when newer ones are unavailable. Needs a dedicated error return type and to let the caller decide how to handle the error.

joestringer commented on pkg/datapath/linux/probes/probes.go:
```diff
@@ -422,3 +395,101 @@ func (p *ProbeManager) KernelConfigAvailable() bool {
+		return err
+	}
+	if err != nil {
+		log.WithError(err).Fatal("failed to probe large instruction limit")
```
If you see the caller checking VTEP support for instance, you'll find that it's an optional feature. The caller already calls `log.Fatal(...)` but in that case the user gets the pointer that if they disable VTEP, they may still be able to successfully run Cilium. Fataling out here deprives the user of that knowledge.

joestringer reviewed (changes requested):
Thanks for working on this! The probe changes in general look good. The fataling out inside the probe library doesn't make sense to me, so I included a few comments on some of them below. I think that the probe library should just return the unsupported error to the caller (wrapped with the type) so that the caller can decide whether the kernel feature is a hard requirement and whether to either fall back to a different implementation or to fatal out itself and provide the user with instructions about how to mitigate the issue (eg, disable a feature).

joestringer reviewed (approved):
You would think that after commenting on the same thing a few times, I would have noticed that I just completely overlooked the exact code I was describing that's already in the code, but evidently I just missed that. I'm not sure what I was thinking. The main feedback I gave was around that and the feedback was invalid, so I'll mark this as approved.

I've looked back over the remaining feedback and it's just this point around defining functions in the asm package which I think isn't quite right in upstream cilium/ebpf library given the lack of guarantee from eBPF runtimes about the meaning for individual BPF_CALL parameter numbers. I'd suggest looking at that a bit more closely in the library and moving them to a package that is runtime specific (ie Linux eBPF as an eBPF runtime target).

tklauser commented on pkg/datapath/linux/probes/probes.go:
```diff
@@ -312,31 +305,11 @@ func (p *ProbeManager) GetOptionalConfig() map[KernelParam]kernelOption {
 }
 
 // GetMapTypes returns information about supported BPF map types.
+// Deprecated. Use HaveMapType() instead.
```
The only remaining user of this method seems to be in `probes_privileged_test.go` for `TestMapTypes`. The purpose of that test seems to be testing `GetMapTypes`. I think we can thus remove the test and this method as well. While at it maybe also add a test for `HaveMapType` instead? Can both be done as a follow-up PR.

ti-mo replied:
> The only remaining user of this method seems to be in probes_privileged_test.go for TestMapTypes. The purpose of that test seems to be testing GetMapTypes. I think we can thus remove the test and this method as well. 

Good point, done!

> While at it maybe also add a test for HaveMapType instead?

Package `features` already has an extensive test suite in the lib itself, so that should be covered. The specific map/prog types the Cilium agent probes for were even used to choose the initial batch of handcrafted tests that are expected to pass on specific kernel versions we need to support!

`HaveMapType` and friends are simple logging wrappers that can't really be tested.

tklauser reviewed (approved):
One non-blocking comment regarding a deprecated method which can be addressed in a follow-up PR. Otherwise LGTM, thanks!

ti-mo commented on pkg/datapath/linux/probes/probes.go:
```diff
@@ -422,3 +395,101 @@ func (p *ProbeManager) KernelConfigAvailable() bool {
+		return err
+	}
+	if err != nil {
+		log.WithError(err).Fatalf("failed to probe MapType: %s", mt.String())
```
Nit: all operands passed to string formatters like `Fatalf()` always get their `String()` methods called, if any. These may be better supplied with `WithField()`, though. I'll push a fixup.
